Fabric particles v1

Sollace opened the pull request:
This is an alternative approach to implementing the particles API. The forward facing interface is a little simpler and registering particles that use a custom sprite supplier is identical to vanilla.

Creating a particle type (client and server):
```
	public static final DefaultParticleType SIMPLE_TEST_PARTICLE = FabricParticleTypes.simple();
	public static final DefaultParticleType CUSTOM_TEST_PARTICLE = FabricParticleTypes.simple();

    @Override
    public void onInitialize() {
        Registry.register(Registry.PARTICLE_TYPE, new Identifier("testmod", "simple"), SIMPLE_TEST_PARTICLE);
        Registry.register(Registry.PARTICLE_TYPE, new Identifier("testmod", "custom"), CUSTOM_TEST_PARTICLE);
    }
```

Registering a particle factory (client):
```
	@Override
	public void onInitializeClient() {
		ParticleFactoryRegistry.getInstance().register(SIMPLE_TEST_PARTICLE, SimpleTestParticle::new);
		ParticleFactoryRegistry.getInstance().register(CUSTOM_TEST_PARTICLE, CustomTestParticle.Factory::new);
	}
```

The particle classes:
```
	@Environment(EnvType.CLIENT)
	static class SimpleTestParticle extends SpriteBillboardParticle {

		public SimpleTestParticle(ParticleEffect effect, World world, double x, double y, double z, double velX, double velY, double velZ) {
			super(world, x, y, z, velX, velY, velZ);

			setSprite(MinecraftClient.getInstance().getItemRenderer().getModels().getSprite(Items.BARRIER));
		}

		@Override
		public ParticleTextureSheet getType() {
			return ParticleTextureSheet.PARTICLE_SHEET_TRANSLUCENT;
		}
	}

	@Environment(EnvType.CLIENT)
	static class CustomTestParticle extends AnimatedParticle {

		protected CustomTestParticle(World world, double x, double y, double z, SpriteProvider sprites) {
			super(world, x, y, z, sprites, 1);

			setSprite(sprites.getSprite(world.random));
		}

		@Environment(EnvType.CLIENT)
		public static class Factory implements ParticleFactory<DefaultParticleType> {

			private final FabricSpriteProvider sprites;

			public Factory(FabricSpriteProvider sprites) {
				this.sprites = sprites;
			}

			@Override
			public Particle createParticle(DefaultParticleType type, World world, double x, double y, double z, double vX, double vY, double vZ) {
			   return new CustomTestParticle(world, x, y, z, sprites);
			}
		}
	}
```

<s>Among these are also some helper. `ParticleUtils` provides some methods to spawn a particle that accept position and velocity as `Vec3d`.</s>

Sollace committed:
Added fabric-particles-v1

Sollace committed:
Create an example mod for creating particles

Sollace committed:
Fix mixin apply error

Sollace committed:
Fix some derps with the test code

Sollace committed:
Rewrite to use raw reflection rather than a gson

kgscialdone commented:
My major concern with this is how much we should abstract over vanilla. The previous PR (#264, mine) uses the existing vanilla methods as much as possible and only provides hooks for things that are inaccessible or extremely unintuitive. While this puts a slightly nicer interface on the API that I could definitely take a couple pointers from, I'm not sure what the verdict is on how much or little we want to abstract, since the policy for existing modules seems to be to leave most things as close to the metal as possible.

Sollace commented:
@swordglowsblue I don't know what you mean by "abstracting". Literally all this does is provide methods to register particles into the vanilla system.

Sollace commented:
ParticleUtils will also just delegate to the method on World.

kgscialdone commented:
The main thing that jumps out at me is particle types - the vanilla method for registering them works just fine, so it's a little odd to me to have an (at least superficially) entirely separate registry for them, especially since Fabric doesn't do anything similar for blocks/items/etc.

```java
Registry.register(Registry.PARTICLE_TYPES, "namespace:id", particleType);
```

I would also suggest that we should still provide a way to access the particle sprite atlas directly, even if SpriteProviders are available (and the reflection makes me a bit squeamish, although it does appear to be necessary in this case unless you just avoid the area entirely like I did).

Sollace commented:
@swordglowsblue The point of providing our own method for this is to require mods to supply an `Identifier` like we do in various other parts of Fabric.

It's a point that's beaten home when I proposed my changes the keybindings: Use `Identifiers` wherever possible.

Also by having our own method this opens us up for when we do eventually want to extend the APIs further.

kgscialdone commented:
> The point of providing our own method for this is to require mods to supply an Identifier like we do in various other parts of Fabric.

Vanilla does not do this. We should not do this over top of vanilla unless there is a separate reason for that abstraction to exist, in which case forcing the use of an `Identifier` is a side bonus (and a completely unnecessary one, since vanilla literally just wraps that string in `new Identifier()` anyway - there is absolutely no difference in function between the two and the `Identifier` constructor takes the exact same string anyways). And in this case, there really isn't much "extending it further" - it's a method to register a particle type. What else is it going to do, ever?

Sollace commented:
@swordglowsblue I argued for the same thing. If you don't like it, go complain to @modmuss50.
![image](https://user-images.githubusercontent.com/6429283/63091717-9428a200-bf5f-11e9-9a66-44ec182ab385.png)

The Keybinding API already forces you to use identifiers. I didn't like that and wanted to remove the restriction, but he made it very clear Fabric wants to use Identifier for everything.

kgscialdone commented:
I agree that Fabric should use identifiers for everything - *everything of our own*. If we're introducing an abstraction over a perfectly functioning vanilla method for no other reason than to force the use of identifiers, then I personally see that as a problem. Forcing identifiers when the abstraction has another legitimate reason to exist is a different story entirely (note that this is the case with the keybinding API, hence I have no particular issue with it there).

Sollace commented:
In the case of the Keybinding api there really isn't a reason. It immediately turns the identifier into a string and throws it away.

Identifiers aren't used, nor does it have any reason they have to be there. If anything Identifiers are _more_ appropriate here because Mojang _does_ use them for particles (just the method takes a string for their own convenience).

It's also important to be explicit here: The register _requires_ identifiers. It needs a namespace and path. You don't want people naming their particles like this:

```Registry.register(Registry.PARTICLE_TYPES, "blue_stone_dust", particleType);```

Your method that accepts strings is unprotected, and it's open to abuse, whether intentional or not. Modders are going to see this and assume they don't need a domain and you'll end up with mods conflicting as they all register their particles in the default namespace.

kgscialdone commented:
That is a problem that vanilla has, which I see no particular reason to go out of our way to fix by introducing a (visually) separate registry system. The method in question isn't mine, it's Mojang's - take it up with them. As is, I find it ill advised to introduce a wrapper around the existing registry methods that serves no purpose other than to force the use of identifiers, even if that leaves an opportunity open for modders to shoot themselves in the foot; if they're going to create that problem for themselves, they're already doing it with blocks and items anyway.

kgscialdone commented:
Secondly, the issue with your logic here: 

> Your method that accepts strings is unprotected, and it's open to abuse, whether intentional or not. Modders are going to see this and assume they don't need a domain and you'll end up with mods conflicting as they all register their particles in the default namespace.

Guess what the identifier constructor does? Oh yeah, it takes that exact same string. You *can* pass them explicitly as two separate strings, but if you only pass one string, it parses it as an ID, and if that ID doesn't have a namespace, it defaults to `minecraft`. Exactly the same as the vanilla methods that take strings directly. Again, if they're going to create that problem for themselves, they're already doing it one way or another - this is a documentation and training issue, not an issue that can or necessarily should be solved entirely by API design.

Sollace commented:
> this is a documentation and training issue, not an issue that can or necessarily should be solved entirely by API design.

What documentation? People don't read documentation. Documents aren't always visible.

Have you never heard of light and dark patterns? Baking the correct usage into the API design is literally _the best_ way of doing this.

Yes, they can create an identifier with the default namespace, but at least there it has to be conscious decision. They're _willingly_ shooting themselves in the foot.

I'm also not saying they shouldn't be able to do it, there are legitimate uses where people might want to replace an existing vanilla particle, but the apis should clearly reflect what values you're expected to give.

As for
> "extending it further" - it's a method to register a particle type. What else is it going to do, ever?

The default ones can be changed to return a `FabricParticleType` with added functionality, if anyone decides to do so.

It also does the job of creating the particle types that have protected constructors that you'd otherwise have to make your own anon classes.

Sollace committed:
We don't need to stringify the identifier

kgscialdone commented:
> What documentation? People don't read documentation. Documents aren't always visible. Have you never heard of light and dark patterns? Baking the correct usage into the API design is literally the best way of doing this.

Apparently you haven't, since you misused the term. A dark pattern is an intentional decision to draw attention away from the option you don't want people to take and towards the option that *they don't want to take*, not leaving them enough rope to hang themselves by. Baking in the correct usage is good, but if the API serves no other purpose than to layer that check over top of the existing vanilla code, it might as well not exist.

> Yes, they can create an identifier with the default namespace, but at least there it has to be conscious decision. They're willingly shooting themselves in the foot.
> 
> I'm also not saying they shouldn't be able to do it, there are legitimate uses where people might want to replace an existing vanilla particle, but the apis should clearly reflect what values you're expected to give.

Which is why vanilla has both overloads - the API is clear from a simple glance at the autocomplete options or source code. If they don't understand how identifiers work, they're going to screw it up anyway, and if they do understand how identifiers work, raw strings won't cause them any undue trouble.

> It also does the job of creating the particle types that have protected constructors that you'd otherwise have to make your own anon classes.

This is part of why I made the architecture decisions I did in my implementation - I provide factory methods for particle types since they're fairly unintuitive, but I leave registry to the modder via the existing vanilla method instead of arbitrarily and unnecessarily abstracting it.

Sollace commented:
> A dark pattern is an intentional decision to draw attention away from the option you don't want people to take and towards the option that they don't want to take,

What we're implementing here is a light pattern. Look it up.

> This is part of why I made the architecture decisions I did in my implementation - I provide factory methods for particle types since they're fairly unintuitive, but I leave registry to the modder via the existing vanilla method instead of arbitrarily and unnecessarily abstracting it.

And this is exactly the decision from your api that I _don't_ like. If you're adding factory methods, they should _register the type itself_. I don't want to have to make three registry calls to add a custom particle, and sit wondering why the game is crashing only to realise that my particle types _need to be registered_, and the method to register them _isn't_ the same method you used to create them and to add the particle factories.

kgscialdone commented:
> If you're adding factory methods, they should register the type itself.

That goes against the basic principles of the factory pattern. Factory methods create. Nothing more.

> I don't want to have to make three registry calls to add a custom particle, and sit wondering why the game is crashing only to realise that my particle types need to be registered, 

That's how everything else in the game works. Blocks, items, entities. The list goes on.

> and the method to register them isn't the same method you used to create them and to add the particle factories.

Of course it's not, because that goes against the basic principles of the factory pattern.

The way you've done it here isn't bad. I just personally disagree with it, because it doesn't follow the established patterns of either Fabric or Vanilla. Unnecessary abstractions in some places, erroneous merging of functions in others, reflection underlying an API that can be designed around it by changing the problem space a bit, and overall a less fitting and cohesive addition when taken together with the API as a whole in my opinion. And that's all it is - my opinion.

grondag commented:
> @swordglowsblue I argued for the same thing. If you don't like it, go complain to @modmuss50.
> ![image](https://user-images.githubusercontent.com/6429283/63091717-9428a200-bf5f-11e9-9a66-44ec182ab385.png)
> 
> The Keybinding API already forces you to use identifiers. I didn't like that and wanted to remove the restriction, but he made it very clear Fabric wants to use Identifier for everything.

I believe the lesson there was "don't make breaking changes to the API outside of a major version cycle," not "use identifiers for everything."

Sollace commented:
@grondag Yeah, which I later ended up cprrecting. Still waiting on feedback on the new PR btw.

@swordglowsblue I'm sick of arguing with you on this. TL;DR I designed it the way it is for a reason, _several_ reasons. If you don't like it, don't use it.

JamiesWhiteShirt commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/impl/particles/ParticleFactoryRegistryImpl.java:
```diff
@@ -0,0 +1,96 @@
+    static final class AccessHack {
+        private static Constructor<?> constr;
+
+        private static Constructor<?> getConstructor() throws ReflectiveOperationException {
```
Have you considered forName using loader's MappingResolver API with the intermediary name of the class we're trying to access `net/minecraft/class_702$class_4090`? I suspect this could break in some dumb and surprising way if the ParticleManager class ever changes.

Sollace replied:
I haven't, I wasn't aware that existed. I'd considered using a mixin to add an interface onto the type as some way to more easily identify that but if there's a proper API that would help a lot.

Do I need to depend on a specific module to use it?

JamiesWhiteShirt commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/api/particles/ParticleUtils.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Utility class for spawning particles.
+ */
+public final class ParticleUtils {
```
Opinion: This isn't the kind of utility that Fabric API should expose. Fabric API should be lightweight and this adds very minimal value.

Sollace replied:
What? You mean this isn't lightweight enough!?

JamiesWhiteShirt replied:
The existence of this class "adds weight" so to speak. It's far less necessary than everything else in the API.

Sollace replied:
Fair enough. I _can_ remove it. It was added more because it was part of the particle code I had before, and I felt it added value to be able to work more with vectors.

JamiesWhiteShirt commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/impl/particles/ParticleFactoryRegistryImpl.java:
```diff
@@ -0,0 +1,96 @@
+            try {
+                return (SpriteProvider)getConstructor().newInstance(MinecraftClient.getInstance().particleManager);
+            } catch (ReflectiveOperationException e) {
+                e.printStackTrace();
```
```suggestion
                throw new IllegalStateException("net.minecraft.client.particle.ParticleManager.SimpleSpriteProvider.<init>() is gone!", e);
```

Sollace replied:
Oop! I totally forgot to remove that before committing.

grondag commented:
While I appreciate the intent of this PR (though the timing could have been much better) I favor the PR that is already in last call.  

I am concerned the reflection hacks used for opening up the vanilla sprite provider mechanism will be apt to break as Mojang continues to refactor their rendering code.  The vanilla approach seems brittle to start with, and the fact they've not exposed it only reinforces that concern.    

Regarding the  helpers and simplified semantics - it will always be somewhat subjective, but IMO the benefits here don't merit the extra maintenance burden.

Sollace committed:
Update fabric-particles-v1/src/main/java/net/fabricmc/fabric/impl/particles/ParticleFactoryRegistryImpl.java

Co-Authored-By: Erlend Åmdal <erlend@aamdal.com>

Sollace commented:
@grondag Yea

> 
> 
> While I appreciate the intent of this PR (though the timing could have been much better) I favor the PR that is already in last call.
> 
> I am concerned the reflection hacks used for opening up the vanilla sprite provider mechanism will be apt to break as Mojang continues to refactor their rendering code. The vanilla approach seems brittle to start with, and the fact they've not exposed it only reinforces that concern.
> 
> Regarding the helpers and simplified semantics - it will always be somewhat subjective, but IMO the benefits here don't merit the extra maintenance burden.

I mentioned before on Discord, though I don't know if anyone saw it: But I believe the reflection could be removed if/when we get shadow classes support in Mixins. Not sure what the status of that feature is, though.

Sollace commented:
Also worth noting: I haven't looked much into the other PR, but it seems like a lot more maintanance to expose custom sprite sheets and reimplement the same features as Vanilla than to allow people to use what's already there.

Sollace committed:
Bad commit

Sollace committed:
Use the mapping resolver to simplify this

mattmess1221 commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/impl/particles/ParticleFactoryRegistryImpl.java:
```diff
@@ -72,16 +73,14 @@ public void injectValues(Int2ObjectMap<ParticleFactory<?>> factories, Map<Identi
-                if (!cls.isInterface() && SpriteProvider.class.isAssignableFrom(cls)) {
-                    constr = cls.getDeclaredConstructor(ParticleManager.class);
-                    constr.setAccessible(true);
+            String className = FabricLoader.getInstance().getMappingResolver().mapClassName("intermediary", "net.minecraft.class_702$class_4090");
```
You should add a comment so you know exactly what this is without needing to look it up.

Sollace replied:
`throw new IllegalStateException("net.minecraft.client.particle.ParticleManager.SimpleSpriteProvider.<init>() is gone!", e);`

Sollace replied:
`Uses reflection to obtain new instances of the private inner class ParticleManager.SimpleSpriteProvider.`

mattmess1221 commented:
> If you don't like it, don't use it.

That's not how this works. You're doing a PR because you want other people to use this. If they don't like it and don't use it, what's the point of merging? I'd understand if this was a standalone project, but you're trying to make this a standard in fabric.

Sollace commented:
@killjoy1221 As far as I'm concerned, he's complaining about having to use Identifiers which have been standard to Fabric for a long time now.

Sollace committed:
Remove ParticleUtils

kgscialdone commented:
No, I am not. My complaint has nothing to do with identifiers and everything to do with unnecessary abstraction that achieves nothing, alongside some other already stated issues with the core architecture of your PR. If you're going to reduce my responses to little more than childish whining about boilerplate, then kindly close your PR and do it in the privacy of your own brain instead of being petty and passive aggressive about it here.

At the end of the day we both just want what's best for the project here. Try not to forget that.

Sollace committed:
Nerf the ParticleTypeRegistry

Sollace committed:
Removed the need for reflection, and added a way to access the sprite sheet directly

Sollace commented:
Updated my PR! Managed to remove _all_ the reflection and also added a method to access the sprite sheet.

Now when you register a particle using the factory type you will get an instance of `FabricSpriteProvider`. That works basically identically to the vanilla `SpriteProvider` class however it has two extra methods:

```
public interface FabricSpriteProvider extends SpriteProvider {

    /**
     * Returns the entire particles texture atlas.
     */
    SpriteAtlasTexture getAtlas();

    /**
     * Gets the list of all sprites available for this particle to use.
     * This is defined in your resourcepack.
     */
    List<Sprite> getSprites();
}
```

kgscialdone commented on fabric-particles-v1/build.gradle:
```diff
@@ -0,0 +1,6 @@
+version = getSubprojectVersion(project, "0.1.1")
+
+dependencies {
+	compile project(path: ':fabric-api-base', configuration: 'dev')
```
This should probably also depend on `fabric-textures-v0`, since `ClientSpriteRegistryCallback` is required to register sprites to the particle atlas for use with particles - even though it's not used directly by the API, this would ensure that it's present for use.

kgscialdone commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/api/particles/FabricParticleTypes.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    // Constructor is (gasp!) protected
+    public static class Simple extends DefaultParticleType {
```
This and `Complex` seem completely unnecessary - why not just use anonymous classes here, since it's hidden behind utility methods anyway?

Sollace replied:
I don't much like Anon classes because I always feel like they're going to have some kind of performance impact over staticaly-defined ones. Having the named classes here also means they're available for anyone who wishes to use them.

Pyrofab replied:
Anonymous classes are strictly the same as named ones, they are generated at compile time.

Sollace replied:
What about situations like this?

```
static Cow createCow(String sound) {
   return new Cow() {
       @Override
       public String moo() { return sound; }
   }
}
```
(and also lambda closurs)? I don't know whether that causes it to generate a new class at every invocation.

kgscialdone replied:
It doesn't. It will generate a new *instance* at each invocation, as expected, with that instance having its own copy of `sound`.

kgscialdone commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/impl/particles/FabricParticleManager.java:
```diff
@@ -0,0 +1,118 @@
+
+    class FabricSpriteProviderImpl implements FabricSpriteProvider {
+
+        private List<Identifier> picks;
```
`picks` seems like a strange name for this concept - perhaps `spriteIDs` or `pendingSprites` might be better?

Sollace replied:
It's an internal name, so... *shrug*

Picks to me is synonymous with the particle's preferred selection. I'm just holding it so the actual list of sprites can be lazily resolved.

Sollace replied:
Decided to just make this change. `spriteIds` sounds appropriate enough.

kgscialdone commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/mixin/client/particles/MixinParticleManager.java:
```diff
@@ -0,0 +1,65 @@
+        fabricParticleManager.injectValues();
+    }
+
+    @Redirect(method = "method_18836(Lnet/minecraft/client/resource/ResourceManager;Lnet/minecraft/util/Identifier;Ljava/util/Map;)V",
```
This should probably not be a redirect, since it may create compatibility-related crashes (not as badly as an overwrite, but still enough to be a concern). If reasonably possible, I'd replace it with an injection instead.

Sollace replied:
This is a bit tricky. For this to work I need to have a reference to `ParticleTextureData` (a local) _and_ the parameters to the parent method.

I can do a ~~injectino~~ injection here, but I'd be missing the local, or I can inject into `ParticleTextureData` but I'd be missing the contextual values from here.

Redirect conveniently seems to be the only method that provides both.

Sollace replied:
Relating to this, I have a possible solution to get rid of the redirect. However I'm tired and can't test it right now, so I'll get back to this tomorrow and push it if everything still works.

Sollace replied:
Tested and pushed! It now uses an Inject at HEAD and if our system knows of a particle with a matching ID, we do the de-serializing and cancel the upstream method.

kgscialdone commented on fabric-testmods/java/net/fabricmc/fabric/particles/ParticleModClient.java:
```diff
@@ -0,0 +1,97 @@
+import net.minecraft.util.registry.Registry;
+import net.minecraft.world.World;
+
+public class ParticleModClient implements ClientModInitializer, ModInitializer {
```
This and other test mod files should be removed from the PR and used locally only. Don't want to ship them with the API, after all.

Sollace replied:
It's in the fabric-testmods module as an example (something people badgered me about including with my previous PR).

There _is_ a lot more garbage involved with actually getting the test mod to run, that I've avoided committing.

kgscialdone replied:
Ah, okay. My bad, I missed that in the path.

kgscialdone reviewed (changes requested):
This looks really good, better than mine to be honest. There are a few things I'd like to see, but altogether this seems like a very good implementation.

Sollace committed:
Add FabricTexturesV0 4 the little 1s

Pyrofab commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/impl/particles/FabricParticleManager.java:
```diff
@@ -0,0 +1,118 @@
+        return picks;
+    }
+
+    class FabricSpriteProviderImpl implements FabricSpriteProvider {
```
Does this need to be a package-private non-static class ?

Sollace replied:
I guess not? Do you rather I make it private?

Pyrofab replied:
`private static` preferably. Code should always be using the most restrictive access modifier possible.

Sollace replied:
I made it private. It can't be static because that would break it.

Pyrofab replied:
I guess I misread it. I thought it had no reference to the enclosing class.

Pyrofab commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/impl/particles/FabricParticleManager.java:
```diff
@@ -0,0 +1,118 @@
+        });
+    }
+
+    FabricSpriteProviderImpl getProvider(Identifier id) {
```
Why leave this method package-private ?

Pyrofab commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/impl/particles/ParticleFactoryRegistryImpl.java:
```diff
@@ -0,0 +1,51 @@
+
+    final Int2ObjectMap<ParticleFactory<?>> factories = new Int2ObjectOpenHashMap<>();
+    final Int2ObjectMap<PendingParticleFactory<?>> constructors = new Int2ObjectOpenHashMap<>();
+    final Map<Identifier, Integer> constructorsIdsMap = new HashMap<>();
```
Does any of these need to be package-private ?

Sollace replied:
Yeah. They're accessed from `FabricParticleManager`.

Pyrofab commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/api/particles/FabricParticleTypes.java:
```diff
@@ -0,0 +1,82 @@
+/**
+ * Methods for creating particle types, both simple and using an existing attribute factory.
+ */
+public final class FabricParticleTypes {
```
This may need a private constructor.

Pyrofab commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/impl/particles/ParticleFactoryRegistryImpl.java:
```diff
@@ -0,0 +1,51 @@
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+
+public class ParticleFactoryRegistryImpl implements ParticleFactoryRegistry {
```
Could probably add `final` on there.

Pyrofab reviewed (commented):
Seems to be forgetting a few java modifiers

Sollace committed:
imo We don't have to be THIS restrictive but whatever

Sollace committed:
Remove the explicit classes.

Sollace committed:
Forgot that thing

mattmess1221 commented:
There's a lot of client classes in this api. Should some of them go into a client package so it's clear that it may load some client classes?

Sollace commented:
@killjoy1221 Does anyone agree with this? There'd basically be just one class that _isn't_ in a client package.

kgscialdone commented:
I think the intention of killjoy's comment is good, but considering that FabricParticleTypes is the only non-client class, it's a little trickier to say whether that's the right call. Probably best to leave that call to one of the higher ups, I'd say.

Sollace commented:
I agree. It doesn't really feel worth while doing. I feel most people are already aware particles are 99% client-side.

Sollace committed:
picks -> spriteIds

Sollace committed:
Rewrote to remove the @Redirect

liach commented on fabric-testmods/java/net/fabricmc/fabric/particles/ParticleModClient.java:
```diff
@@ -0,0 +1,97 @@
+
+	@Override
+	public void onInitializeClient() {
+		onInitialize();
```
You shouldn't make this call.

Sollace replied:
Hm? Oh that's just there because I couldn't get it to run by itself.

Player3324 commented:
Please use tabs and remove the blank line before the first member declarations.

Player3324 commented:
> @killjoy1221 Does anyone agree with this? There'd basically be just one class that _isn't_ in a client package.

Any module that isn't or won't always be 100% side specific has to be split in client/common/server, even if it is future module reordering would benefit from everything using this structure.

Sollace commented:
@sfPlayer1 
> Please use tabs

I was told that can be fixed when merging. My IDE won't easily let me use tabs for my java files.

> remove the blank line before the first member declarations.

Wait wat. _Why?_

Player3324 commented:
This needs to be fixed first, otherwise the inline review comments may lose their association. Every IDE I am aware of can use tabs.

The style has to match the existing code.

Sollace committed:
Downgrade to tabs

Sollace committed:
Make the test mod non-functional

Sollace commented:
> 
> 
> This needs to be fixed first, otherwise the inline review comments may lose their association. Every IDE I am aware of can use tabs.
> 
> The style has to match the existing code.

They have been switched to tabs. Also removed that call I'd forgotten to remove from the test mod, @liach

Sollace committed:
C O N V E N T I O N S

kgscialdone commented:
Alright, look. Overall this is a great PR and to some extent I get where you're coming from on the conventions - they're imperfect and could be bikeshedded for years before they'd get there, and even then there'd be people who disagree with them. But the passive aggressive commit messages are getting a bit much.

Sollace commented:
> passive aggressive commit messages are getting a bit much.

I like to have fun with my commit messages, and they're a convenient way for me to vent. I'm not going to do something I don't like and pretend that I agree with it.

No doubt they're going to squash this when they merge anyway, like has been done before. In fact I recommend they squash, because regardless of commit messages, all the patches at the end of every PR definitely _does not_ look good.

liach commented:
So which particle api pr will be accepted? This or #264?

kgscialdone commented:
Despite my personal stake in #264, my vote goes to #341 as the more complete of the two.

Mhowser commented:
How many more reviews do we need to get this merged?

Sollace commented:
I feel like the magic number is three approvals, plus a week of buffer time where I don't get any other changes I need to address.

Prospector commented:
There's not really a magic number, but there's literally 0 approvals on this PR so that'll be a nope for now.

liach commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/api/particles/ParticleFactoryRegistry.java:
```diff
@@ -0,0 +1,64 @@
+	 * @param <T> The type of particle effects this factory deals with.
+	 */
+	@FunctionalInterface
+	public interface PendingParticleFactory<T extends ParticleEffect> {
```
No need for extra public

Sollace replied:
I wasn't sure if it was needed or not, so I left it to be safe. Also doesn't Fabric have some requirement for `public` on interface members?

Pyrofab replied:
Should be consistent, either put the redundant public modifier everywhere or nowhere. The codebase tends towards the former.

Sollace replied:
Alright. Should I update my interfaces to add the redundant modifier then?

I don't remember if this has always been the case, or if the modifier being optional is new to java 8+.

Sollace replied:
Updated to include the public for all the interfaces. My IDE removes all redundant modifiers on save so it was... difficult.

Sollace replied:
Also an interesting observation: My IDE deems all the `public`s as redundant in interfaces, except the ones on the inner-class definitions of interfaces.

liach commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/impl/particles/FabricParticleManager.java:
```diff
@@ -0,0 +1,127 @@
+
+		Identifier file = new Identifier(id.getNamespace(), "particles/" + id.getPath() + ".json");
+
+		try (Reader reader = new InputStreamReader(manager.getResource(file).getInputStream(), Charsets.UTF_8)) {
```
`StandardCharsets.UTF_8`

Sollace replied:
This was taken from Mojang's code, so Charsets is what they were using.

liach replied:
Don't use "mojang code" as an excuse. We need to updated code instead of keeping legacy code.

Sollace replied:
Well what's the difference? They're the same constant, and what I'm saying is if Mojang is using it, why can't we? It's going to be part of the environment regardless.

liach commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/mixin/client/particles/MixinParticleManager.java:
```diff
@@ -0,0 +1,63 @@
+@Mixin(ParticleManager.class)
+public abstract class MixinParticleManager implements VanillaParticleManager {
+
+	private final FabricParticleManager fabricParticleManager = new FabricParticleManager(this);
```
Does this even work? I think you are supposed to move the value initialization to an inject to the return of all constructors instead

Sollace replied:
Nope. This works fine. I've been doing it like this in production at least a year now.

liach reviewed (commented):
Other things look all good!

Mhowser reviewed (approved):
Great work,  can't wait to see this in action.

Sollace committed:
I had to fight my IDE to do this. I hope it was worth it.

Sollace committed:
Use the built-in constant instead of google's

Sollace committed:
Separate into a client/common package

modmuss50 reviewed (commented):
Ive taken a quick look over it and it looks good. 

This seems close to being able to put on Last Call

Player3324 commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/api/client/particles/FabricSpriteProvider.java:
```diff
@@ -0,0 +1,36 @@
+import net.minecraft.client.texture.Sprite;
+import net.minecraft.client.texture.SpriteAtlasTexture;
+
+public interface FabricSpriteProvider extends SpriteProvider {
```
Please add a class comment quickly explaining what it is and how it's used (referencing ParticleFactoryRegistry) and reference FabricParticleTypes for the example

Player3324 commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/api/particles/FabricParticleTypes.java:
```diff
@@ -0,0 +1,68 @@
+/**
+ * Methods for creating particle types, both simple and using an existing attribute factory.
+ */
+public final class FabricParticleTypes {
```
Please add an example here similar to the PR's opening post

Player3324 commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/api/client/particles/FabricSpriteProvider.java:
```diff
@@ -0,0 +1,36 @@
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.particles;
```
The package should use singular "particle" instead of "particles" and carry the module api version (v1) after the module name. In this file that'd be net.fabricmc.fabric.api.client.particle.v1

Sollace replied:
> carry the module api version (v1) after the module name. 

Since when has this being a requirement? Including versions in the package automatically guarantees no backwards compatibility.

grondag replied:
Typically v1 and v2 would coexist for some period of time with v1 in a deprecated state. 

Changes that can be backwards compatible ( generally additive changes) would not prompt a major version change.

grondag replied:
Typically v1 and v2 would coexist for some period of time with v1 in a deprecated state. 

Changes that can be backwards compatible ( generally additive changes) would not prompt a major version change.

Sollace replied:
Y'all already have such a long-ass package name, though. Sure, I'll change them. Just have to start up Eclipse.

Player3324 commented on fabric-particles-v1/src/main/resources/assets/testmod/particles/custom.json:
```diff
@@ -0,0 +1,12 @@
+    "minecraft:spell_1",
+    "minecraft:spell_0"
+  ]
+}
```
Some files don't have a trailing new line, please add where missing

Player3324 commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/mixin/client/particles/MixinParticleManager.java:
```diff
@@ -0,0 +1,63 @@
+
+@Mixin(ParticleManager.class)
+public abstract class MixinParticleManager implements VanillaParticleManager {
+
```
Extra blank link

Player3324 commented on fabric-particles-v1/src/main/java/net/fabricmc/fabric/api/client/particles/ParticleFactoryRegistry.java:
```diff
@@ -0,0 +1,64 @@
+
+/**
+ * Registry for adding particle factories on the client.
+ */
```
Please reference FabricParticleTypes for the example here

Player3324 reviewed (commented):
This is almost ready, it's only missing the api package fix, embedded example+refs and a small style fix.

Sollace committed:
Fix formatting issue

Sollace committed:
Fix formatting and move test mod resources out of the module

Sollace committed:
Docs

Sollace committed:
Versionafitize package names

Player3324 reviewed (approved):
The v1 suffix is only the api packages, please adjust accordingly. I'll approve anyway to get the last call countdown going. Thank you very much for all the work!

Sollace committed:
Unversion the implementation classes

Sollace committed:
Update the mixins.json with the correct package name
